allennlp/models/reading_comprehension/dialog_qa.py

```diff
@@ -0,0 +1,407 @@
+    Question Answering in Context (EMNLP 2018) paper for baseline.
+    In this set-up, a single instance is a dialog, list of question answer pairs.
+    Parameters
+    ----------
+    vocab : ``Vocabulary``
+    text_field_embedder : ``TextFieldEmbedder``
+        Used to embed the ``question`` and ``passage`` ``TextFields`` we get as input to the model.
+    phrase_layer : ``Seq2SeqEncoder``
+        The encoder (with its own internal stacking) that we will use in between embedding tokens
+        and doing the bidirectional attention.
+    span_start_encoder : ``Seq2SeqEncoder``
+        The encoder that we will use to incorporate span start predictions into the passage state
+        before predicting span end.
+    span_end_encoder : ``Seq2SeqEncoder``
+        The encoder that we will use to incorporate span end predictions into the passage state.
+    dropout : ``float``, optional (default=0.2)
+        If greater than 0, we will apply dropout with this probability after all encoders (pytorch
+        LSTMs do not apply dropout to their last layer).
+    num_context_answers : ``int``, optional (default=0)
+        If greater than 0, the model will consider previous question answering context.
+    mask_lstms : ``bool``, optional (default=True)
+        If ``False``, we will skip passing the mask to the LSTM layers.  This gives a ~2x speedup,
+        with only a slight performance decrease, if any.  We haven't experimented much with this
+        yet, but have confirmed that we still get very similar performance with much faster
+        training times.  We still use the mask for all softmaxes, but avoid the shuffling that's
+        required when using masking with pytorch LSTMs.
+    """
+
+    def __init__(self, vocab: Vocabulary,
+                 text_field_embedder: TextFieldEmbedder,
+                 phrase_layer: Seq2SeqEncoder,
+                 residual_encoder: Seq2SeqEncoder,
+                 span_start_encoder: Seq2SeqEncoder,
+                 span_end_encoder: Seq2SeqEncoder,
+                 initializer: InitializerApplicator,
+                 dropout: float = 0.2,
+                 num_context_answers: int = 0,
+                 marker_embedding_dim: int = 10,
+                 mask_lstms: bool = True) -> None:
+        super().__init__(vocab)
+        self._num_context_answers = num_context_answers
+        self._text_field_embedder = text_field_embedder
+        self._phrase_layer = phrase_layer
+
+        encoding_dim = phrase_layer.get_output_dim()
+        max_turn_length = 12
+
+        self._matrix_attention = LinearMatrixAttention(encoding_dim, encoding_dim, 'x,y,x*y')
+        self._merge_atten = TimeDistributed(torch.nn.Linear(encoding_dim * 4, encoding_dim))
+
+        self._residual_encoder = residual_encoder
+        self._prev_ans_marker = torch.nn.Embedding((num_context_answers * 4) + 1, marker_embedding_dim)
```

Comment:
I think this belongs in the `if` block below, because this is only actually used if `num_context_answers > 0`.